### DISCUSSION TITLE
Generate sensitivity tuning curve fails

### DISCUSSION THREAD

**Discussion by akandrew332** [angelolab/toffy#233] [Q&A] [answered]
Installed toffy successfully on CAC, however no rosetta_matrices folder created.  Additionally, generating sensitivity curve fails.  Identified good sweep for generating tuning curve.  What exactly is supposed to be copied to the normalization curve folder? Instructions state to copy all FOVs into that folder.  Each voltage in the sweep is a different folder itself.  Tried copying all the different folders for moly scans at different voltages for the sweep into the normalization curve folder.  This failed.  Are only specific files supposed to placed in the normalization curve folder?  Thanks!
![function_fit](https://user-images.githubusercontent.com/19846608/188238277-f5d0ebdc-e643-4f80-bfa7-82f8ad4e4907.jpg)

**Comment by ngreenwald** [comment, marked as answer]
Yes, you should be able to just directly copy each run folder into the overall normalization folder. 

i.e.

```sh
normalization_curve
│ 
├── 20220901_2200V
│    ├── fov-1-scan-1.bin
│    ├── fov-1-scan-1.json
├── 20220901_2225V
│    ├── fov-1-scan-1.bin
│    ├── fov-1-scan-1.json
├── 20220901_2250V
│    ├── fov-1-scan-1.bin
│    ├── fov-1-scan-1.json
```
 
Good catch about the rosetta matrices folder, we'll add that into the notebook

**Reply by akandrew332** [reply]
Okay, tried using the above directory structure and files.  Still getting a bunch of errors.  Generated function_fit is the same.
![Normerror1](https://user-images.githubusercontent.com/19846608/188243093-5537983d-50ac-426b-98d1-bb605cea0de2.JPG)

**Reply by ngreenwald** [reply]
Got it, all those divide by zero and covariance errors are likely coming from the same problem: FOVs with very low counts being included. Can you take a look at the mass spectrum of the FOVs you selected using MIBIControl and make sure there are positive counts for the masses in the 90s? Sometimes the first part of a detector sweep is at such a low voltage that there is essentially no signal.

**Reply by akandrew332** [reply]
Success.  Used generated count csvs to find and remove FOVs with zero counts.  This didn't work.  Then removed one more FOV with low, but non-zero counts.  This worked.  Thanks and have a great weekend.
![Success](https://user-images.githubusercontent.com/19846608/188244961-043816d3-37e1-4321-9bf0-a83cf2e04ba8.JPG)

**Reply by ngreenwald** [reply]
No problem. We'll add in a check to automatically remove FOVs with low counts in the future so this doesn't happen

**Reply by akandrew332** [reply]
That would be really great! I've been looking at the generated plot and the shape of it looks quite different from the example in toffy.  Additionally, the values on the x-axis are quite different (both the range and magnitude).  Any insights as to why for any of the above?  I've pasted the Toffy example curve for comparison to the one generated from the detector sweep.  Additionally, why wouldn't you rerun a tuning curve when the detector is replaced?
![ToffyExampleCurve](https://user-images.githubusercontent.com/19846608/188688461-232fe107-0cdb-481c-99d2-53593b6640b9.JPG)

**Reply by ngreenwald** [reply]
In addition to removing the low end, we also clip the super high voltage FOVs from the top end when constructing the curve. We'll add this is as a default as well

**Reply by akandrew332** [reply]
I've been trying to figure out what that x axis is actually representing.  Under the example curve in the notebook, it indicates Detector Gain.  Judging by the numbers it appears that it is more likely to be pulse height.  Corresponds better to the csvs of pulse height generated in each of the detector gain setting FOV folders from the sweep.

**Reply by ngreenwald** [reply]
@camisowers will you also add axis labels

**Reply by camisowers** [reply]
Hi @akandrew332, we're working on incorporating an automatic check for extreme count values, to avoid the low count issue you encountered. Would you mind providing an example of one of your channel counts files that is on the higher end? This would help us understand how different machines are running; we're looking to find an upper threshold to our count range as well.

**Reply by akandrew332** [reply]
No problem.  Do you just need one of the folders at the higher voltage range, or do you want the entire sweep with all the different voltages?

**Reply by camisowers** [reply]
If you could just attach a few of the channel_counts and pulse_heights csv files for higher voltages that were generated for the tuning curve calculation that would be great!

**Reply by akandrew332** [reply]
File attached.
[pulseheightandcounts.zip](https://github.com/angelolab/toffy/files/9602746/pulseheightandcounts.zip)

**Reply by camisowers** [reply]
Thank you!

**Reply by akandrew332** [reply]
No problem!  By the way, those wider range plots might have some value...  Here is our current detector:
![newest detector](https://user-images.githubusercontent.com/19846608/191132299-991df7f4-92cb-4c4b-b64d-1dfbd365b907.jpg)

**Reply by akandrew332** [reply]
Here's a previous faulty one:
![Faulty June](https://user-images.githubusercontent.com/19846608/191132336-8be20946-921e-4817-ba34-2ad9d43000c4.jpg)

**Reply by akandrew332** [reply]
Should we be doing normalization curves every time the detector is replaced?

**Reply by ngreenwald** [reply]
The outliers between 10k and 20k (on the x-axis) are what makes the plot look different. If you exclude those, you can imagine that all curves have an elbow point around 8k (on the x-axis), which is what we're looking for, that the shape of the curve is the same. We don't know for sure, but so far haven't seen significant differences between detectors.
